File: context/poc/agent/agent/tools/shell.py
import subprocess

from agents import function_tool


@function_tool
def run_shell(cmd: str) -> str:
    """Execute a shell command and return the output."""
    try:
        result = subprocess.run(
            cmd, shell=True, capture_output=True, text=True, timeout=30
        )
        output = ""
        if result.stdout:
            output += result.stdout
        if result.stderr:
            output += f"\nSTDERR:\n{result.stderr}"
        return output or "(no output)"
    except subprocess.TimeoutExpired:
        return "Error: Command timed out after 30 seconds"
    except Exception as e:
        return f"Error executing command: {e}"

File: example_agents/adk_js/src/tools/shell.ts
import { execSync } from 'node:child_process';
import { FunctionTool } from '@google/adk';
import { z } from 'zod';

export const runShell = new FunctionTool({
  name: 'run_shell',
  description: 'Execute a shell command and return the output.',
  parameters: z.object({
    cmd: z.string().describe('Shell command to execute'),
  }),
  execute({ cmd }) {
    try {
      const stdout = execSync(cmd, { timeout: 30_000, encoding: 'utf-8' });
      return stdout || '(no output)';
    } catch (e: unknown) {
      const err = e as { killed?: boolean; stderr?: string; message?: string };
      if (err.killed) return { error: 'Command timed out after 30 seconds' };
      if (err.stderr) return { error: err.stderr };
      return { error: err.message };
    }
  },
});
